alf/drivers/policy_driver.py

```diff
@@ -0,0 +1,257 @@
+
+    def get_initial_time_step(self):
+        """Returns the initial action_time_step."""
+        time_step = self.env.current_time_step()
+        action = common.zero_tensor_from_nested_spec(self.env.action_spec(),
+                                                     self.env.batch_size)
+        return make_action_time_step(time_step, action)
+
+    def algorithm_step(self, time_step, state):
+        if self._observation_transformer is not None:
+            time_step = time_step._replace(
+                observation=self._observation_transformer(time_step.
+                                                          observation))
+        return self._algorithm_step(time_step, state)
+
+    def run(self, max_num_steps=None, time_step=None, policy_state=None):
+        """Take steps in the environment for max_num_steps.
+
+        If in training mode, algorithm.train_step() and
+        algorithm.train_complete() will be called.
+        If not in training mode, algorith.predict() will be called.
+
+        The observers will also be called for every environment step.
+
+        Args:
+            max_num_steps (int): stops after so many environment steps. Is the
+                total number of steps from all the individual environment in
+                the bached enviroments including StepType.LAST steps.
+            time_step (ActionTimeStep): optional initial time_step. If None, it
+                will use self.get_initial_time_step(). Elements should be shape
+                [batch_size, ...].
+            policy_state (nested Tensor): optional initial state for the policy.
+
+        Returns:
+            time_step (ActionTimeStep): named tuple with final observation,
+                reward, etc.
+            policy_state (nested Tensor): final step policy state.
+        """
+        if time_step is None:
+            time_step = self.get_initial_time_step()
+        if policy_state is None:
+            policy_state = self._initial_state
+        return self._run(
+            time_step=time_step,
+            policy_state=policy_state,
+            max_num_steps=max_num_steps)
+
+    @abc.abstractmethod
+    def _algorithm_step(self, time_step, state):
+        """Return policy_step for this time_step and state
+        """
+        pass
+
+    @abc.abstractmethod
+    def _run(self, max_num_steps, time_step, policy_state):
+        """Take steps in the environment for max_num_steps."""
+        pass
+
+    def predict(self, max_num_steps, time_step, policy_state):
+        maximum_iterations = math.ceil(max_num_steps / self._env.batch_size)
+        [time_step, policy_state] = tf.while_loop(
+            cond=lambda *_: True,
+            body=self._eval_loop_body,
+            loop_vars=[time_step, policy_state],
+            maximum_iterations=maximum_iterations,
+            back_prop=False,
+            name="predict_loop")
+        return time_step, policy_state
+
+    def summary(self, loss_info, grads_and_vars):
```

Comment:
summary => _summary because it is not supposed to be called by other classes
and make a short comment.